Fix for ANR on release (at least one case)

stevemayhew opened the pull request:
This fixes an ANR we are seeing on Player.release().  The problem is the `MediaCodec` listener callbacks hold a lock that blocks any other message to `MediaCodec`'s handler.  This can cause a call into `MediaCodec` to not complete

The bug number is: https://github.com/google/ExoPlayer/issues/4352

stevemayhew commented on library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java:
```diff
@@ -1750,7 +1753,8 @@ public void onFrameRendered(MediaCodec codec, long presentationTimeUs, long nano
-      onProcessedTunneledBuffer(presentationTimeUs);
+
+      handler.post(() -> onProcessedTunneledBuffer(presentationTimeUs));
     }
```
Might be good for MediaCodecRenderer to have an actual Handler that we can post messages too, in the event there are other things like this...   Note, some concern this could queue lots of messages as well.

The Handler is of course on the player's thread so it shares the message queue with `ExoPlayerImplInternal`

Everything else but the code on line 938 above is safe to do in `onProcessTunnelBuffer()` in this same handler `dispatchMessage()`.   It's the calls back into `MediaCodec` that need to be avoided (deferred to a subsequent `Message` that doesn't hold MediaCodec locks.

tonihei commented:
Kevin, could you have a look at this PR to see if it makes sense?

krocard commented on library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java:
```diff
@@ -1740,8 +1740,11 @@ public CodecMaxValues(int width, int height, int inputSize) {
   @TargetApi(23)
   private final class OnFrameRenderedListenerV23 implements MediaCodec.OnFrameRenderedListener {
 
+    protected Handler handler;
```
private final

tobyworks commented:
Just leaving a comment here because i want to follow this PR. Currently we are being bothered with some sporadic ANR's. And i very well believe this could fix it.

krocard commented on library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java:
```diff
@@ -1750,7 +1753,8 @@ public void onFrameRendered(MediaCodec codec, long presentationTimeUs, long nano
         // Stale event.
         return;
       }
```
This stall event detection needs to be done in the lambda.

krocard replied:
You have marked this done, have you have not change that code. Have you forgotten to upload a patch?

stevemayhew replied:
@krocard I've put the check for listener changed (stale event), but also cleaned up the handler to avoid leaks and prevent the stale event from ever happening...  If you agree we can remove the stale check, but kind of hard to ever test.

krocard commented on library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java:
```diff
@@ -1750,7 +1753,8 @@ public void onFrameRendered(MediaCodec codec, long presentationTimeUs, long nano
       }
-      onProcessedTunneledBuffer(presentationTimeUs);
+
+      handler.post(() -> onProcessedTunneledBuffer(presentationTimeUs));
```
A comment should be added that the codec is not reentrant from a callback.

tonihei replied:
This is also fixed in the platform code now. So this workaround can be behind an "if(Util.SDK_INT < 30)" to avoid the post on upcoming Android versions.

krocard replied:
For reference, it was fixed by https://android-review.googlesource.com/1156807.

krocard reviewed (changes requested):
The CL needs to be rebased.

stevemayhew commented:
That’s validating...  can add the link to that in a comment so we can
remove the workaround when the older versions of Android finally age our

On Fri, Nov 8, 2019 at 5:45 AM Kevin Rocard <notifications@github.com>
wrote:

> *@krocard* commented on this pull request.
> ------------------------------
>
> In
> library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java
> <https://github.com/google/ExoPlayer/pull/6603#discussion_r344179271>:
>
> > @@ -1750,7 +1753,8 @@ public void onFrameRendered(MediaCodec codec, long presentationTimeUs, long nano
>          // Stale event.
>          return;
>        }
> -      onProcessedTunneledBuffer(presentationTimeUs);
> +
> +      handler.post(() -> onProcessedTunneledBuffer(presentationTimeUs));
>
> For reference, it was fixed by
> https://android-review.googlesource.com/1156807.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/ExoPlayer/pull/6603?email_source=notifications&email_token=AADBF6ASP4DG2ORD5YHNM4TQSVURLA5CNFSM4JHCNHG2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCK5G37A#discussion_r344179271>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADBF6CVYJUV2K2P42MOSWDQSVURLANCNFSM4JHCNHGQ>
> .
>

stevemayhew committed:
Cleanup tunneled onFrameRenderListener

Cleanup the handler to prevent stale messages and potential handler leaks

stevemayhew commented on library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java:
```diff
@@ -506,6 +506,9 @@ protected void onDisabled() {
+      tunnelingOnFrameRenderedListener.destroyHandler(getCodec());
+    }
     tunnelingOnFrameRenderedListener = null;
     try {
```
Explicitly cleanup the previous handler, and any messages it might have that keep references to the handler itself and implicitly to the outer class.

stevemayhew commented on library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java:
```diff
@@ -1740,8 +1749,23 @@ public CodecMaxValues(int width, int height, int inputSize) {
+        codec.setOnFrameRenderedListener(null, null);
+      }
+      handler.removeCallbacksAndMessages(null);
     }
```
I added in the stale check, however this code will prevent you from ever getting messages after the handler reference is dropped (assuming you call it first).

stevemayhew commented on library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java:
```diff
@@ -1750,7 +1774,24 @@ public void onFrameRendered(MediaCodec codec, long presentationTimeUs, long nano
+      //
+      // The work around simply queues the processing to a subsequent message, where
+      // the lock will not be held.
+      //
+      if (Util.SDK_INT < 30) {
+        handler.post(() -> {
+          if (this == tunnelingOnFrameRenderedListener) {   // event not stale
```
Here's the stale detect, and comment explaining why we have to do this at all.  I'll rebase in the next commit and pick up the changes from 5579cc89c

stevemayhew reviewed (commented):
We are running with this change in our released code now, will of course let you know of any problems we have.  We have stopped seeing the ANR :-)

bjitivo commented on library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java:
```diff
@@ -513,6 +513,9 @@ protected void onDisabled() {
     clearReportedVideoSize();
     clearRenderedFirstFrame();
     frameReleaseTimeHelper.disable();
+    if (tunnelingOnFrameRenderedListener != null && getCodec() != null) {
```
Probably should call destroyHandler even if codec is null ...

bjitivo commented on library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java:
```diff
@@ -1172,6 +1178,9 @@ private void clearRenderedFirstFrame() {
       MediaCodec codec = getCodec();
       // If codec is null then the listener will be instantiated in configureCodec.
       if (codec != null) {
+        if (tunnelingOnFrameRenderedListener != null) {
```
Have to do the destroy unconditionally to protect against a queued render that will run and see the new listener.

bjitivo replied:
I mean do the destroy whether or not there is a codec.

stevemayhew replied:
Thanks Bryan, made this fix in commit 43700ea

stevemayhew committed:
Cleanup onFrameRenderListener handler...

Even if the codec is null (presume it's handler was run down first, this happens in `MediaCodec.stop()`) make sure to run down our own Handler.

krocard commented on library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java:
```diff
@@ -1172,6 +1178,9 @@ private void clearRenderedFirstFrame() {
+        if (tunnelingOnFrameRenderedListener != null) {
+          tunnelingOnFrameRenderedListener.destroyHandler(codec);
+        }
         tunnelingOnFrameRenderedListener = new OnFrameRenderedListenerV23(codec);
```
If there is a listener installed, this leads to 2 calls to calls to `codec.setOnFrameRenderedListener`. One with null, the next one with the new listener.
This leads to disabling codec timestamp reporting then reenabling it immediately. Which is a costly operation as far as I can tell.
Thus I think that destoryHandler should not setOnFrameRenderedListener.
How about creating a function to install a new Listener:
```
MediaCodec.OnFrameRenderedListener oldListener = tunnelingOnFrameRenderedListener;
tunnelingOnFrameRenderedListener = new OnFrameRenderedListenerV23(codec);
if (oldListener != null) {
  oldListener.destroyHandler()
}
```
Maybe add a OnFrameRenderedListenerV23.disable method to codec.setOnFrameRenderedListener(null, null); in disable(). Although I'm not sure it is necessary.

krocard replied:
If you prefer, I can merge your PR as is and do a patch with this optimization over it.

krocard replied:
@stevemayhew, let me know if you would prefer do this change yourself or I can go along and do it on top of yours. We would like to merge your PR. Thanks

stevemayhew replied:
Removing the old handler is just an optimization.  The check for stale events is sufficient.  I will add the function as you describe and the disable() function but not call it for now with a nite with your comments as to why.  Should have this today for you to review

stevemayhew replied:
This commit 72f2e7d4fa6a29aa0cf6bc3a1bf2a8738b4f8ac0 adds the method and removes the double call to `codec.setOnFrameRenderedListener` if the codec is not being disabled.

stevemayhew committed:
Fix not to call codec.setOnFrameRenderedListener(null, null) unless stoping codec

Refator to lift a method to create (and destroy) the OnFrameRenderedListenerV23() and split destroying the Handler from removing the MediaCodec onFrameRenderedListener to avoid double call to codec.setOnFrameRenderedListener()

koAlech commented:
We're also getting these ANRs in production. Any estimate when this PR be merged?

krocard commented:
Sorry for the delay, I am working on it right now. A few changes were suggested by the team but this PR will be part of the next minor release.
Note however that this change only solves a deadlock in tunneling mode during `OnFrameRenderedListener.onFrameRendered` calls.
Other deadlocks have been observed during `Player.release()` and are not solved by this PR.

stevemayhew commented:
Pretty special edge case solved by this pull, you must be running codec in
tunneling mode to see it.

Best to provide a bugreport with the ANR. One of the ExoPlayer devs might
no best if the signature is unique then


On Sun, Jan 12, 2020 at 11:42 PM Alik Hochner <notifications@github.com>
wrote:

> We're also getting these ANRs in production. Any estimate when this PR be
> merged?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/ExoPlayer/pull/6603?email_source=notifications&email_token=AADBF6D3YO5LNPF6RXF5HQ3Q5QLOBA5CNFSM4JHCNHG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIXY4PQ#issuecomment-573541950>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADBF6ELYAHPNO3PGFK4EKDQ5QLOBANCNFSM4JHCNHGQ>
> .
>
